gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/CommitActivityImpl.java

```diff
@@ -0,0 +1,176 @@
+      commitTaskStates(jobState, taskStateQueue, globalGobblinContext);
+      return taskStateQueue.size();
+    } catch (Exception e) {
+      //TODO: IMPROVE GRANULARITY OF RETRIES
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          "Failed to commit dataset state for some dataset(s) of job <jobStub>",
+          IOException.class.toString(),
+          new IOException(e),
+          null
+      );
+    }
+  }
+
+  /**
+   * Commit task states to the dataset state store.
+   * @param jobState
+   * @param taskStates
+   * @param jobContext
+   * @throws IOException
+   */
+  void commitTaskStates(State jobState, Collection<TaskState> taskStates, JobContext jobContext) throws IOException {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = createDatasetStatesByUrns(taskStates);
+    final boolean shouldCommitDataInJob = JobContext.shouldCommitDataInJob(jobState);
+    final DeliverySemantics deliverySemantics = DeliverySemantics.AT_LEAST_ONCE;
+    //TODO: Make this configurable
+    final int numCommitThreads = DEFAULT_NUM_COMMIT_THREADS;
+    if (!shouldCommitDataInJob) {
+      log.info("Job will not commit data since data are committed by tasks.");
+    }
+
+    try {
+      if (!datasetStatesByUrns.isEmpty()) {
+        log.info("Persisting {} dataset urns.", datasetStatesByUrns.size());
+      }
+
+      List<Either<Void, ExecutionException>> result = new IteratorExecutor<>(Iterables
+          .transform(datasetStatesByUrns.entrySet(),
+              new Function<Map.Entry<String, JobState.DatasetState>, Callable<Void>>() {
+                @Nullable
+                @Override
+                public Callable<Void> apply(final Map.Entry<String, JobState.DatasetState> entry) {
+                  return new SafeDatasetCommit(shouldCommitDataInJob, false, deliverySemantics, entry.getKey(),
+                      entry.getValue(), false, jobContext);
+                }
+              }).iterator(), numCommitThreads,
+          ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("Commit-thread-%d")))
+          .executeAndGetResults();
+
+      IteratorExecutor.logFailures(result, null, 10);
+
+      if (!IteratorExecutor.verifyAllSuccessful(result)) {
+        // TODO: propagate cause of failure
+        String jobName = jobState.getProperties().getProperty(ConfigurationKeys.JOB_NAME_KEY, "<job_name_stub>");
+        throw new IOException("Failed to commit dataset state for some dataset(s) of job " + jobName);
+      }
+    } catch (InterruptedException exc) {
+      throw new IOException(exc);
+    }
+  }
+
+  /**
+   * Organize task states by dataset urns.
+   * @param taskStates
+   * @return
+   */
+  public static Map<String, JobState.DatasetState> createDatasetStatesByUrns(Collection<TaskState> taskStates) {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = Maps.newHashMap();
+
+    //TODO: handle skipped tasks?
```

Comment:
Gobblin has a semantic for skipped tasks where workunits can be skipped but their state is saved, not sure why/when it would occur, it's not very clear and lacks documentation here. `workunit.skip` is the config